File: src/api/rules.ts
import { invoke } from "@tauri-apps/api/core";
import type { ApplyResult, EvaluationResult } from "../types";

export async function validateRules(yaml: string): Promise<void> {
  return invoke("validate_rules", { yaml });
}

export async function previewRules(yaml: string): Promise<EvaluationResult> {
  return invoke("preview_rules", { yaml });
}

export async function applyRules(yaml: string): Promise<ApplyResult> {
  return invoke("apply_rules", { yaml });
}

export async function readTextFile(path: string): Promise<string> {
  return invoke("read_text_file", { path });
}

export async function writeTextFile(
  path: string,
  contents: string,
): Promise<void> {
  return invoke("write_text_file", { path, contents });
}

File: src-tauri/src/commands/rules.rs
use tauri::AppHandle;

use crate::commands::library::open_db;
use crate::playlist_rules::{self, ApplyResult, EvaluationResult};

#[tauri::command]
pub fn validate_rules(yaml: String) -> Result<(), String> {
    playlist_rules::validate_rules(&yaml)
}

#[tauri::command]
pub fn preview_rules(app: AppHandle, yaml: String) -> Result<EvaluationResult, String> {
    let db = open_db(&app)?;
    playlist_rules::preview_rules(&db, &yaml)
}

#[tauri::command]
pub fn apply_rules(app: AppHandle, yaml: String) -> Result<ApplyResult, String> {
    let db = open_db(&app)?;
    playlist_rules::apply_rules(&db, &yaml)
}

#[tauri::command]
pub fn read_text_file(path: String) -> Result<String, String> {
    std::fs::read_to_string(&path).map_err(|e| format!("Failed to read {}: {}", path, e))
}

#[tauri::command]
pub fn write_text_file(path: String, contents: String) -> Result<(), String> {
    std::fs::write(&path, contents).map_err(|e| format!("Failed to write {}: {}", path, e))
}
